groups: support hierarchical groups.yaml files

nikhita opened the pull request:
For https://github.com/kubernetes/k8s.io/issues/460

Requiring all groups be added to a single `groups.yaml` can become
tedious to maintain. This commit allows splitting groups into
hierarchical `groups.yaml` files, akin to OWNERS files.

This commit contains the following changes:

- We now hardcode files containing groups config to have the file name
`groups.yaml`.

- We no longer allow the group config to be specified at an URL.

- The `groups-file` field in config.yaml is renamed to `groups-path`.
`groups-path` is the directory containing groups.yaml files. It must be
an absolute path. If not specified, it defaults to the directory
containing `config.yaml`.

- `readGroupsConfig` uses `groups-path` as the root dir and recursively
walks through all dirs and files. When it sees a `groups.yaml` file,
it reads the config and adds its groups to `config.Groups`.

- `config.Groups` finally contains the merged groups config from all
`groups.yaml` files.

---

This is WIP because:

- As https://github.com/kubernetes/k8s.io/issues/928#issuecomment-640993935 mentions, we probably need documented policies about how we'd split groups.yaml before enabling this. I just wanted to get this PR out to get the ball rolling.

- I moved `community@kubernetes.io` to the `sig-contributor-experience` dir and seeded OWNERS with the root OWNERS in k/community as an example, we'd probably need to tweak OWNERS/groups a bit more.

/hold
to ensure we have policies in place before enabling this

/assign @cblecker @dims @spiffxp

nikhita commented:
cc @mrbobbytables @castrojo 
might be useful if we want to include more gsuite accounts (xref https://github.com/kubernetes/steering/issues/166)

dims commented:
/assign @spiffxp @cblecker

spiffxp commented on groups/reconcile.go:
```diff
@@ -202,30 +202,66 @@ func readConfig(configFilePath string, confirmChanges bool) error {
+// recursively walks through all dirs and files. It reads the GroupsConfig
+// from all files that have the name groupsFileName and adds the groups in
+// each GroupsConfig to config.Groups.
+func readMultipleGroupsConfig(configFilePath string, groupsFileName string, config *GroupsConfig) error {
+	rootDir := filepath.Dir(configFilePath)
```
nit: I think it would be cleaner if this function took `rootDir` directly, and the work of deriving it was moved elsewhere

spiffxp commented on groups/reconcile.go:
```diff
@@ -202,30 +202,66 @@ func readConfig(configFilePath string, confirmChanges bool) error {
+	})
+}
+
+func readGroupsConfigFromURL(groupsURL *url.URL, config *GroupsConfig) error {
```
Perhaps we should drop this functionality entirely.  It's not exercised and it's not tested.

nikhita replied:
> Perhaps we should drop this functionality entirely. It's not exercised and it's not tested.

Removed.

spiffxp commented on groups/reconcile.go:
```diff
@@ -202,30 +202,66 @@ func readConfig(configFilePath string, confirmChanges bool) error {
 }
 
-func readGroupsConfig(configFilePath string, groupsConfigFilePath string, config *GroupsConfig) error {
+func readGroupsConfig(configFilePath string, groupsFileName string, config *GroupsConfig) error {
```
Capitalization nit
```suggestion
func readGroupsConfig(configFilePath string, groupsFilename string, config *GroupsConfig) error {
```

As with `readMultipleGroupsConfig` I think this should take an explicit dir instead of passing along `configFilePath` and deriving from it later

spiffxp commented on groups/groups_test.go:
```diff
@@ -25,25 +25,59 @@ import (
 var cfg GroupsConfig
 
-var groupsConfigPath = flag.String("groups-config", "./groups.yaml", "Path to groups config")
+var groupsConfigFileName = flag.String("groups-file", "groups.yaml", "Filename containing groups config")
```
My personal preference would be to alter things to take an explicit `groups-path` or `groups-dir`, instead of auto-deriving it from the location of `config.yaml`

I also wonder if it would be useful to allow files named something other than `groups.yaml`, in the same way that prow jobs can be defined in arbitrarily named files.  Not blocking on this, just thinking out loud. I could foresee wg-k8s-infra using this to have a rbac file, a staging groups file, etc.

nikhita replied:
> explicit groups-path or groups-dir, instead of auto-deriving it from the location of config.yaml

> I also wonder if it would be useful to allow files named something other than groups.yaml

This is interesting! Would something like "read all yaml files under the directory specified by `groups-dir`" make sense? (I still need to look at how prow does it today)

thockin commented:
Why?  Having a strong one-file-per-directory convention is nice because it
aligns with OWNERS

On Mon, Jul 6, 2020 at 9:41 AM Nikhita Raghunath <notifications@github.com>
wrote:

> *@nikhita* commented on this pull request.
> ------------------------------
>
> In groups/groups_test.go
> <https://github.com/kubernetes/k8s.io/pull/996#discussion_r450347998>:
>
> >  )
>
>  var cfg GroupsConfig
>
> -var groupsConfigPath = flag.String("groups-config", "./groups.yaml", "Path to groups config")
> +var groupsConfigFileName = flag.String("groups-file", "groups.yaml", "Filename containing groups config")
>
> explicit groups-path or groups-dir, instead of auto-deriving it from the
> location of config.yaml
>
> I also wonder if it would be useful to allow files named something other
> than groups.yaml
>
> This is interesting! Would something like "read all yaml files under the
> directory specified by groups-dir" make sense? (I still need to look at
> how prow does it today)
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/k8s.io/pull/996#discussion_r450347998>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVFCIFZFJVCKXELX4Y3R2H5DFANCNFSM4OQQSBBQ>
> .
>

thockin commented:
We may even want to link directory-name to group-name to lock it own to
allow OWNERS delegation more

On Mon, Jul 6, 2020 at 9:50 AM Tim Hockin <thockin@google.com> wrote:

> Why?  Having a strong one-file-per-directory convention is nice because it
> aligns with OWNERS
>
> On Mon, Jul 6, 2020 at 9:41 AM Nikhita Raghunath <notifications@github.com>
> wrote:
>
>> *@nikhita* commented on this pull request.
>> ------------------------------
>>
>> In groups/groups_test.go
>> <https://github.com/kubernetes/k8s.io/pull/996#discussion_r450347998>:
>>
>> >  )
>>
>>  var cfg GroupsConfig
>>
>> -var groupsConfigPath = flag.String("groups-config", "./groups.yaml", "Path to groups config")
>> +var groupsConfigFileName = flag.String("groups-file", "groups.yaml", "Filename containing groups config")
>>
>> explicit groups-path or groups-dir, instead of auto-deriving it from the
>> location of config.yaml
>>
>> I also wonder if it would be useful to allow files named something other
>> than groups.yaml
>>
>> This is interesting! Would something like "read all yaml files under the
>> directory specified by groups-dir" make sense? (I still need to look at
>> how prow does it today)
>>
>> —
>> You are receiving this because your review was requested.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/k8s.io/pull/996#discussion_r450347998>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABKWAVFCIFZFJVCKXELX4Y3R2H5DFANCNFSM4OQQSBBQ>
>> .
>>
>

nikhita commented:
Addressed comments. PTAL. The updates are documented in the commit message and the issue body.

TLDR:

- We now hardcode files containing groups config to have the file name
`groups.yaml`.

- We no longer allow the group config to be specified at an URL.

- The `groups-file` field in config.yaml is renamed to `groups-path`.
`groups-path` is the directory containing groups.yaml files. It must be
an absolute path. If not specified, it defaults to the directory
containing `config.yaml`.

nikhita commented:
> We may even want to link directory-name to group-name to lock it own to
allow OWNERS delegation more

I haven't added a validation for this because we might want to create directories based on sig names. I've added discussing the policy for splitting groups.yaml to [this week's wg-k8s-infra meeting](https://docs.google.com/document/d/16VBfsFMynA7tObzuZGPpw-sKDKfFc_T5W_E4IeEIaOQ/edit#).

spiffxp commented on groups/sig-contributor-experience/OWNERS:
```diff
@@ -0,0 +1,10 @@
+# See the OWNERS docs at https://go.k8s.io/owners
+
+approvers:
```
```suggestion
reviewers:
```
I would happily merge this PR as-is if this wasn't present.  I do want to open up the pool but we're not there yet.

If the goal is to fast-track membership changes, I would suggest we test against a known list of groups that is under /groups/OWNERS approval only

If the goal is to allow sigs/wgs/whatever to create whatever @kubernetes.io groups they want, I would at a minimum require adherence to common naming conventions and settings

nikhita replied:
Should I just removed the contribex dir and comment out the test for it? I just meant it as an example that this PR works :woman_shrugging: 

I agree that we shouldn't merge this as-is.

spiffxp replied:
Leaving the dir and everything is fine (and preferred!), if the approvers are changed to reviewers for now

nikhita replied:
Done!

spiffxp commented:
/approve
/lgtm

bartsmykla commented:
lgtm

nikhita commented:
Discussed in today's wg k8s infra meeting, let's merge and iterate :sparkles: 

We discussed a [few approaches towards policy](https://docs.google.com/document/d/16VBfsFMynA7tObzuZGPpw-sKDKfFc_T5W_E4IeEIaOQ/edit), and decided to go with whatever is easiest to setup validation for. I'll take a dab at that over the weekend :)

/hold cancel
